src/rtp/srtp.rs

```diff
@@ -36,71 +111,83 @@ pub struct SrtpContext {
 }
 
 impl SrtpContext {
-    pub fn new(srtp_key: SrtpKey) -> Self {
-        let (rtp, rtcp) = Derived::from_key(&srtp_key);
-
-        SrtpContext {
-            rtp,
-            rtcp,
-            srtcp_index: 0,
-        }
-    }
-
-    // SRTP layout
-    // [header, [rtp, (padding + pad_count)], hmac]
-
-    //     0                   1                   2                   3
-    //     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+<+
-    //    |V=2|P|X|  CC   |M|     PT      |       sequence number         | |
-    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //    |                           timestamp                           | |
-    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //    |           synchronization source (SSRC) identifier            | |
-    //    +=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+ |
-    //    |            contributing source (CSRC) identifiers             | |
-    //    |                               ....                            | |
-    //    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //    |                   RTP extension (OPTIONAL)                    | |
-    //  +>+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //  | |                          payload  ...                         | |
-    //  | |                               +-------------------------------+ |
-    //  | |                               | RTP padding   | RTP pad count | |
-    //  +>+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+<+
-    //  | ~                     SRTP MKI (OPTIONAL)                       ~ |
-    //  | +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //  | :                 authentication tag (RECOMMENDED)              : |
-    //  | +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ |
-    //  |                                                                   |
-    //  +- Encrypted Portion*                      Authenticated Portion ---+
-
     pub fn protect_rtp(
         &mut self,
         buf: &[u8],
         header: &RtpHeader,
         srtp_index: u64, // same as ext_seq
     ) -> Vec<u8> {
-        let iv = self.rtp.salt.rtp_iv(*header.ssrc, srtp_index);
-
         let hlen = header.header_len;
         let input = &buf[hlen..];
+
         assert!(
             input.len() % SRTP_BLOCK_SIZE == 0,
             "RTP body should be padded to 16 byte block size, {header:?} with body length {} was not", input.len()
         );
 
-        let mut output = vec![0_u8; buf.len() + SRTP_HMAC_LEN];
-        self.rtp
-            .enc
-            .encrypt(&iv, input, &mut output[hlen..])
-            .expect("rtp encrypt");
+        match &mut self.rtp {
```

Comment:
These match statements in `protect_rtp`, `unprotect_rtp`, `protect_rtcp`, and `unprotect_rtcp` could be removed by expanding the cipher specific API so the code here could be common

Reply:
Yeah, why isn't it? :)

Reply:
Because that's part of making it beautiful, and I haven't gotten that far. I wanted to get a feel for the different profiles before going down that route.

Happy to address it another PR or in this one

Reply:
Let's do it in the next one.